### ISSUE TITLE
feat(nexmark source): generate events in a single source

### ISSUE THREAD

**Issue by fuyufjh** [risingwavelabs/risingwave#6747]
> > The dimension tables in Nexmark generator should have fewer records. But as we poll the sources in a fair manner, the actual result is that, there are 15-50x dimension records generated, compared to the bid in the same time. This makes the workload not realistic. Will this lead to problems?
>
> It's indeed a potential problem that could lead to benchmark results that are far from the real cases. 😇
>
> _Originally posted by @fuyufjh in https://github.com/risingwavelabs/risingwave/issues/6571#issuecomment-1331721743_

In short, the NexMark source doesn't give off events of 3 tables in our expected ratio (person:aution:bid = 1:3:46 by default). Instead, the events are generated in near 1:1:1, because they are all polled at similar frequencies.

Flink's NexMark source combines these 3 tables into 1 single source and then separate them with filter conditions in 3 views. Well, I finally understand why they did so. In this way, the portion of 3 kinds of events are ensured to be of the expected ratio.

https://github.com/nexmark/nexmark/blob/8ff9866ef7908967a7c40a82dd51c27140f738e4/nexmark-flink/src/main/resources/queries/ddl_gen.sql

https://github.com/nexmark/nexmark/blob/8ff9866ef7908967a7c40a82dd51c27140f738e4/nexmark-flink/src/main/resources/queries/ddl_gen.sql

**Comment by lmatz** [comment]
A similar problem was encountered in the Demo before: https://github.com/risingwavelabs/risingwave-demo/pull/33

Separated generation of data for two input streams will lead to a weird workload, i.e. events A happen after events B while in reality, this can never happen.
Because we pre-gen the data, then it all depends on the speed of the source to determine whether A or B happens first.

**Comment by BugenZhao** [comment]
Before optimization in #6419 and #6460, we always generate all kinds of events in each `Generator` and drop the unused ones, which is a kind of "proof of work" (😄) and should give an expected event num ratio if the CPU time is the same. However, this is not optimal since if we really want events from all three sources, we'll spend 300% CPU time for 100% of events.

So I think "avoid generating unused events" is a must-do, while we may try to reuse the single queue for three kinds?

**Comment by wangrunji0408** [comment]
The incorrect ratio problem could be avoided by specifying a **non-zero and large enough** value for `nexmark.min.event.gap.in.ns`, in order to generate events at a rate that doesn't exceed the downstream processing capabilities. But it seems that most of the time we set it to 0 value. 🥹

**Comment by fuyufjh** [comment]
> So I think "avoid generating unused events" is a must-do, while we may try to reuse the single queue for three kinds?

Since it's a distributed system and the 3 tables may have different partition key, you cannot just do it internally. 

I think the only way to realize it is my proposal.

**Comment by fuyufjh** [comment]
> The incorrect ratio problem could be avoided by setting a **non-zero and large enough** value for `nexmark.min.event.gap.in.ns`, in order to generate events at a rate that doesn't exceed the downstream processing capabilities. But it seems that most of the time we set it to 0. 🥹

Actually I am using a non-zero value (1000ns). But it's very hard to get a proper value, because our target is to benchmark i.e. get the highest performance as much as possible

**Comment by wangrunji0408** [comment]
> > The incorrect ratio problem could be avoided by setting a **non-zero and large enough** value for `nexmark.min.event.gap.in.ns`, in order to generate events at a rate that doesn't exceed the downstream processing capabilities. But it seems that most of the time we set it to 0. 🥹
> 
> Actually I am using a non-zero value (1000ns), but it still need to be less to keep the right "ratio". See, it's very hard to get a proper value, because our target is to benchmark i.e. get the highest performance as much as possible

Yes, in this case different event sources will eventually get out-of-sync, as long as they are independent from each other. So I agree with you that a single source might be the only solution.

**Comment by BugenZhao** [comment]
We may use the materialized source for now.

**Comment by fuyufjh** [comment]
> We may use the materialized source for now.

Yes, that's right! But our major motivation is to do benchmark so materialized source may be a too heavy

**Comment by lmatz** [comment]
Close by #6800, 
throughput: https://github.com/risingwavelabs/risingwave/pull/6800#issuecomment-1343973066.
